tests/e2e/vmc_upgrade_tkg.go

```diff
@@ -0,0 +1,275 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	storagev1 "k8s.io/api/storage/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	fnodes "k8s.io/kubernetes/test/e2e/framework/node"
+	fss "k8s.io/kubernetes/test/e2e/framework/statefulset"
+)
+
+var _ = ginkgo.Describe("Upgrade TKG", func() {
+	f := framework.NewDefaultFramework("vmc-upgrade-tkg")
+	var (
+		client            clientset.Interface
+		namespace         string
+		svNamespace       string
+		scParameters      map[string]string
+		storagePolicyName string
+		manifestPath      = "tests/e2e/testing-manifests/statefulset/nginx"
+		vmcUser           string
+		flag              bool
+	)
+
+	ginkgo.BeforeEach(func() {
+		flag = false
+		client = f.ClientSet
+		namespace = getNamespaceToRunTests(f)
+		svcClient, svNamespace = getSvcClientAndNamespace()
+		nodeList, err := fnodes.GetReadySchedulableNodes(f.ClientSet)
+		framework.ExpectNoError(err, "Unable to find ready and schedulable Node")
+		storagePolicyName = GetAndExpectStringEnvVar(envStoragePolicyNameForSharedDatastores)
+		if !(len(nodeList.Items) > 0) {
+			framework.Failf("Unable to find ready and schedulable Node")
+		}
+		bootstrap()
+		setResourceQuota(svcClient, svNamespace, rqLimit)
+		scParameters = make(map[string]string)
+		storagePolicyName = GetAndExpectStringEnvVar(envStoragePolicyNameForSharedDatastores)
+	})
+	ginkgo.AfterEach(func() {
+		svcClient, svNamespace := getSvcClientAndNamespace()
+		setResourceQuota(svcClient, svNamespace, defaultrqLimit)
+	})
+
+	/*
+		Test to Scale TKG worker nodes
+		Steps
+			1. Create 3 Statefulset pods, PVC with the above created SC and validate PVC is in bound phase
+			2. Verify volumes are attached to the nodes
+			3. Upgrade the TKG cluster
+			4. Scale up the Statefulset pods to 8
+			5. Verify volumes are attached to the nodes
+			6. Wait for some time to scale up operations to complete
+			7. Verify PVC and POD should be in bound and running state
+			8. Scale down the sts pods to 0
+			9. Verify volumes are detached from the nodes
+			10. Delete the statefulsets
+	*/
+
+	ginkgo.It("[upgrade-tkg-vmc] Upgrade TKG", func() {
+		var sc *storagev1.StorageClass
+		var err error
+		var wcpToken string
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		vmcUser = os.Getenv("VMC_USER")
+		if vmcUser == "" {
+			ginkgo.Skip("Env vmcUser is missing")
+		}
+		tkg_image := os.Getenv("TKG_UPGRADE_IMAGE")
+		if tkg_image == "" {
+			ginkgo.Skip("Env TKG_UPGRADE_IMAGE is missing")
+		}
+		tkg_cluster := os.Getenv("TKG_CLUSTER")
```

Comment:
couple of options here
1. change TKG_CLUSTER ot TKG_CLUSTER_FOR_UPGRADE
2. we can create a tkg in the test right, just fetch two params for old image and new image and use the old image to deploy tkg

Reply:
1. Done
2. I don't understand the second comment, here we are not creating tkg, we are updating the already created tkg.
we have separate tests to create tkg, maybe those tests can be enhanced a bit.